nvbench/iteration_space_base.cuh

```diff
@@ -0,0 +1,129 @@
+/*
+ *  Copyright 2022 NVIDIA Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 with the LLVM exception
+ *  (the "License"); you may not use this file except in compliance with
+ *  the License.
+ *
+ *  You may obtain a copy of the License at
+ *
+ *      http://llvm.org/foundation/relicensing/LICENSE.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#pragma once
+
+#include <nvbench/detail/axes_iterator.cuh>
+
+namespace nvbench
+{
+
+/*!
+ * Base class for all axi and axes iteration spaces.
+ *
+ * If we consider an axi to be a container of values, iteration_spaces
+ * would be how we can create iterators over that container.
+ *
+ * With that in mind we get the following mapping:
+ * * linear_axis_space is equivalant to a forward iterator.
+ *
+ * * zip_axis_space is equivalant to a zip iterator.
+ *
+ * * user_axis_space is equivalant to a transform iterator.
+ *
+ * We don't immediately construct the iterators as the active elements,
+ * name, etc can be changed before execution. This class allows for
+ * the deferred iterator creation while keeping the meta data insyc.
+ *
+ */
+struct iteration_space_base
+{
+  using axes_type = std::vector<std::unique_ptr<nvbench::axis_base>>;
+  using axes_info = std::vector<detail::axis_index>;
+
+  using AdvanceSignature =
+    nvbench::detail::axis_space_iterator::AdvanceSignature;
+  using UpdateSignature = nvbench::detail::axis_space_iterator::UpdateSignature;
+
+  /*!
+   * Construct a new derived iteration_space
+   *
+   * @param[input_indices]
+   * @param[output_indices]
+   */
+  iteration_space_base(std::vector<std::size_t> input_indices,
+                       std::vector<std::size_t> output_indices);
```

Comment:
It's not obvious what the `input_indices/output_indices` represent here.

Reply:
I have updated the docs with more information, does that help clarify it?